quote-book/synchronized/src/lib.rs

```diff
@@ -1,111 +1,392 @@
 // Copyright (c) 2023 MobileCoin Inc.
-
-use deqs_quote_book_api::{Error, Pair, Quote, QuoteBook, QuoteId};
+// Copyright (c) 2023 MobileCoin Inc.
+use deqs_quote_book_api::{Error as QuoteBookError, Pair, Quote, QuoteBook, QuoteId};
 use mc_blockchain_types::BlockIndex;
 use mc_crypto_ring_signature::KeyImage;
-use mc_ledger_db::Ledger;
+use mc_ledger_db::{Error as LedgerError, Ledger};
 use mc_transaction_extra::SignedContingentInput;
-use std::ops::RangeBounds;
 
+use std::{
+    ops::RangeBounds,
+    sync::{
+        atomic::{AtomicBool, AtomicU64, Ordering},
+        Arc, Mutex,
+    },
+    thread::{Builder as ThreadBuilder, JoinHandle},
+    time::Duration,
+};
+
+use mc_common::logger::{log, Logger};
 /// A wrapper for a quote book implementation that syncs quotes with the ledger
-#[derive(Clone)]
-pub struct SynchronizedQuoteBook<Q: QuoteBook, L: Ledger + Clone + 'static> {
+pub struct SynchronizedQuoteBookImpl<Q: QuoteBook, L: Ledger + Clone + Sync + 'static> {
```

Comment:
Do you actually need the `Impl` version? Everything here but the join handle is already cloneable, so can the join handle be placed in an Arc Mutex?

Reply:
Thank you for this suggestion.

Reply:
@eranrund @wjuan-mob i'm worried that the "wrap join handle in arc mutex instead of impl" is going to cause a bug, here's what i'm worried about.

1. The reason the QuoteBook requires clone is that, we want to think of it as a handle, and handles to the quote book are passed around by the server.
2. However, after the change where `SynchronizedQuoteBook` implements `QuoteBook` directly, cloning it is not really a handle.
3. Putting the `JoinHandle` in `Arc<Mutex` but nothing else, means that if the `SynchronizedQuoteBook::clone` is ever called, and any clone of `SynchronizedQuoteBook` is dropped, then when it is dropped, `.stop` will be called, and the thread will be joined. Then the remaining clones will be broken because they won't have a thread anymore.

Reply:
lmk if you think i'm missing something

Reply:
maybe its okay if the `QuoteBook` inside the synchronized quotebook is itself wrapped in `Arc` ?

Reply:
willing to be convinced that we should just think of everything in this object as a handle, and then derive clone, but let's please document this.

Reply:
I see, that's a good point.
Maybe the solution is to only execute the stopping and thread waiting when the reference count is 1? https://doc.rust-lang.org/std/sync/struct.Arc.html#method.strong_count can be used to check for that

Reply:
A better solution is probably move everything related to the thread into its own object.
Then `SynchronizedQuoteBook` will only contain two fields (I think): the underlying `QuoteBook` and `Arc<SyncThread>`. `SyncThread` will also hold a clone of the underlying `QuoteBook`, and implement `Drop` to stop the thread and wait for it.

Reply:
Am I correct in understanding that the solution you're suggesting is that I have:

- SynchronizedQuotebook, which implements that logic around filtering quote as they come in. and an arc which contains the SyncThread
- SyncThread which holds the joinhandle and stop_requested for the dbfetcher thread
- DBFetcher thread, which is the same.

Is this substantially different from the impl approach I was using before which was an arc which contained the synchronizedquotebookimpl?

Reply:
fwiw i think the version with the impl approach and an arc was fine, also just implementing `QuoteBook` for `Arc<SynchronizedQuoteBook>` seemed fine to me

Reply:
fwiw, my preference would have been to make `QuoteBook` not imply `Clone`, and whoever is using `QuoteBook` trait and needs to clone it just holds an `Arc<dyn QuoteBook>`, that feels more natural to me. but i'm also okay with the way it is

Reply:
it seems to me like the important parts of what it means to be a quote book are that you can add and remove quotes, search for quotes, etc., and the Clone thing is just sort of orthogonal and has to do with how the server happens to want to use instances of QuoteBook

Reply:
but like, it's a larger change than the scope of this PR to do that

Reply:
> Is this substantially different from the impl approach I was using before which was an arc which contained the synchronizedquotebookimpl?

I agree with this comment, I don't see the difference between this and the sync thread object

Reply:
Implementing the trait on the arc doesn't seem to work. It ends up giving me an error about `only traits defined in the current crate can be implemented for types defined outside of the crate
define and implement a trait or new type instead` I'm not sure if there's a way around that besides using something similar to the impl type approach.

Reply:
right

Reply:
I think eran's approach also makes sense

Reply:
where it's approximately like:

```
struct SyncQB<QB> {
quote_book: QB,
thread: Arc<ThreadState>,
}

struct ThreadState {
    highest_processed_block_index: Arc<AtomicU64>,
    join_handle: Option<JoinHandle<()>>,
    stop_requested: Arc<AtomicBool>,
}
```

Reply:
if that would avoid implement `QuoteBook` on an `impl` object separately from the "main" object then that seems like a good idea

Reply:
or like, it seems like it splits some of the concerns of this object in a way that might be helpful

Reply:
i think i didn't get it at first, but now i see the logic of this

Reply:
Thank you for helping me understand why this is better. I hope that I have implemented your suggestion appropriately.